File: frontend_tests/fx-spec.test.js
// JavaScript half of the shared currency-conversion contract, reading the same
// tests/fixtures/fx-spec.json as the Python and Kotlin tests.
//
// The web has always converted; this pins it so the other two cannot drift away from it again.
import test from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";

import {
  configureMetrics,
  findCurrencyConversionRate,
  normalizeFxRates
} from "../frontend/metrics.js";

const SPEC_PATH = path.resolve(process.cwd(), "tests", "fixtures", "fx-spec.json");

// frontend/metrics.js takes its collaborators through configureMetrics. Of the six, the currency
// helpers touch only toNum, and every rate in the fixture is a plain number — so this minimal
// coercion is equivalent to the app's for these inputs. Deliberately not a copy of app.js's toNum:
// that one parses thousand separators and decimal commas, has its own coverage, and duplicating it
// here would be a fourth implementation of the thing this fixture exists to prevent.
const unusedOutsideCurrencyHelpers = () => {
  throw new Error("The currency helpers should not need this collaborator.");
};

configureMetrics({
  toNum: (value) => {
    const numeric = Number(value);
    return Number.isFinite(numeric) ? numeric : 0;
  },
  findById: unusedOutsideCurrencyHelpers,
  lookupName: unusedOutsideCurrencyHelpers,
  sum: unusedOutsideCurrencyHelpers,
  todayIso: unusedOutsideCurrencyHelpers,
  getState: unusedOutsideCurrencyHelpers
});

function loadSpec() {
  return JSON.parse(fs.readFileSync(SPEC_PATH, "utf8"));
}

function ratesFor(testCase, spec) {
  if ("ratesJson" in testCase) {
    return testCase.ratesJson;
  }
  return testCase.rates || spec.rates;
}

function applicable(spec) {
  return spec.cases.filter((testCase) =>
    (testCase.appliesTo || ["python", "js", "kotlin"]).includes("js")
  );
}

test("the shared FX spec has cases for this implementation", () => {
  assert.ok(applicable(loadSpec()).length > 0, "The shared FX spec is empty.");
});

test("conversion rates match the shared spec", () => {
  const spec = loadSpec();
  const failures = [];

  for (const testCase of applicable(spec)) {
    const actual = findCurrencyConversionRate(testCase.from, testCase.to, ratesFor(testCase, spec));
    if (Math.abs(actual - testCase.expectRate) > spec.tolerance) {
      failures.push(
        `${testCase.id}: rate ${actual}, spec says ${testCase.expectRate} — ${testCase.why}`
      );
    }
  }

  assert.deepEqual(failures, [], `Rates disagree with the shared spec:\n  ${failures.join("\n  ")}`);
});

test("a missing rate never zeroes a holding", () => {
  const spec = loadSpec();

  for (const testCase of applicable(spec)) {
    if (testCase.expectRate !== 0) {
      continue;
    }
    const rate = findCurrencyConversionRate(testCase.from, testCase.to, ratesFor(testCase, spec));
    assert.equal(rate, 0, `${testCase.id}: expected no usable rate — ${testCase.why}`);
  }
});

test("corrupt rate payloads are dropped", () => {
  for (const payload of [{ "USD/PLN": -1 }, "not-json", null, ["list"], { NOTAPAIR: 4 }]) {
    assert.deepEqual(
      normalizeFxRates(payload),
      {},
      `${JSON.stringify(payload)} should normalise to an empty rate table`
    );
  }
});

File: android-app/tax/src/test/kotlin/pl/prywatnyportfel/tax/FxSpecTest.kt
package pl.prywatnyportfel.tax

import java.io.File
import org.json.JSONObject
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * Kotlin half of the shared currency-conversion contract, reading the same
 * tests/fixtures/fx-spec.json as the Python and JavaScript tests.
 *
 * Cases marked appliesTo without "kotlin" are skipped: this module takes rates as a map because it
 * has no JSON parser in its main source set, so the string-payload case belongs to the other two.
 */
class FxSpecTest {

    private fun specFile(): File {
        var dir: File? = File(System.getProperty("user.dir"))
        while (dir != null) {
            val candidate = File(dir, "tests/fixtures/fx-spec.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile
        }
        fail("tests/fixtures/fx-spec.json not found above ${System.getProperty("user.dir")}")
    }

    private fun spec(): JSONObject = JSONObject(specFile().readText())

    private fun toMap(json: JSONObject): Map<String, Any?> =
        json.keys().asSequence().associateWith { key -> json.get(key) }

    private fun ratesFor(case: JSONObject, spec: JSONObject): Map<String, Any?> =
        toMap(case.optJSONObject("rates") ?: spec.getJSONObject("rates"))

    private fun appliesToKotlin(case: JSONObject): Boolean {
        val list = case.optJSONArray("appliesTo") ?: return true
        return (0 until list.length()).any { list.getString(it) == "kotlin" }
    }

    private fun applicableCases(spec: JSONObject): List<JSONObject> {
        val cases = spec.getJSONArray("cases")
        return (0 until cases.length())
            .map { cases.getJSONObject(it) }
            .filter { appliesToKotlin(it) }
    }

    @Test
    fun `spec has cases for this implementation`() {
        assertTrue(applicableCases(spec()).isNotEmpty(), "The shared FX spec is empty.")
    }

    @Test
    fun `conversion rates match the spec`() {
        val spec = spec()
        val tolerance = spec.optDouble("tolerance", 1e-6)
        val failures = mutableListOf<String>()

        for (case in applicableCases(spec)) {
            val actual = FxConversion.findCurrencyConversionRate(
                case.getString("from"),
                case.getString("to"),
                ratesFor(case, spec),
            )
            val expected = case.getDouble("expectRate")
            if (abs(actual - expected) > tolerance) {
                failures.add("${case.getString("id")}: rate $actual, spec says $expected — ${case.getString("why")}")
            }
        }
        assertTrue(failures.isEmpty(), "Rates disagree with the shared spec:\n  " + failures.joinToString("\n  "))
    }

    @Test
    fun `converted amounts match the spec`() {
        val spec = spec()
        val tolerance = spec.optDouble("tolerance", 1e-6)
        val failures = mutableListOf<String>()

        for (case in applicableCases(spec)) {
            val actual = FxConversion.convertCurrency(
                case.get("amount"),
                case.getString("from"),
                case.getString("to"),
                ratesFor(case, spec),
            )
            val expected = case.getDouble("expectConverted")
            if (abs(actual - expected) > tolerance) {
                failures.add("${case.getString("id")}: got $actual, spec says $expected — ${case.getString("why")}")
            }
        }
        assertTrue(failures.isEmpty(), "Conversions disagree with the shared spec:\n  " + failures.joinToString("\n  "))
    }

    @Test
    fun `a missing rate never zeroes a holding`() {
        val spec = spec()
        for (case in applicableCases(spec)) {
            if (case.getDouble("expectRate") != 0.0) continue
            val amount = case.getDouble("amount")
            assertEquals(
                amount,
                FxConversion.convertCurrency(
                    case.get("amount"),
                    case.getString("from"),
                    case.getString("to"),
                    ratesFor(case, spec),
                ),
                "An unusable rate must leave the amount untouched, not zero it.",
            )
        }
    }

    @Test
    fun `corrupt rate payloads are dropped`() {
        assertEquals(emptyMap(), FxConversion.normalizeFxRates(mapOf("USD/PLN" to -1)))
        assertEquals(emptyMap(), FxConversion.normalizeFxRates(mapOf("NOTAPAIR" to 4.0)))
        assertEquals(emptyMap(), FxConversion.normalizeFxRates(null))
        assertEquals(emptyMap(), FxConversion.normalizeFxRates("not-a-map"))
        assertEquals(emptyMap(), FxConversion.normalizeFxRates(listOf("list")))
    }
}
